psi4/driver/driver.py

```diff
@@ -1931,37 +1931,121 @@ def gdma(wfn, datafile=""):
+      A warning will be printed if a theory module is not supported.
+    * Caution! For orbital-optimized correlated methods (e.g. DCT, OMP2) the 'Orbital Energy' field contains ambiguous data.
 
     :examples:
 
     >>> # [1] FCHK file for DFT calculation
     >>> E, wfn = energy('b3lyp', return_wfn=True)
     >>> fchk(wfn, 'mycalc.fchk')
 
+    >>> # [2] FCHK file for correlated densities
+    >>> E, wfn = gradient('ccsd', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk')
+
     """
-    fw = core.FCHKWriter(wfn)
-    fw.write(filename)
+    # * Known limitations and notes *
+    #
+    # OCC: (occ theory module only, not dfocc) is turned off as it densities are not set. 
+    # DFMP2: Contains natural orbitals in wfn.C() and wfn.epsilon() data.
+    # This is fixed to contain respective HF data.
+
+    allowed = ['DFMP2', 'SCF', 'CCENERGY', 'DCT', 'DFOCC']
+    module_ = wfn.module().upper()
+    if (module_ not in allowed):
+        core.print_out(f"FCHKWriter: Theory module {module_} is currently not supported by the FCHK writer.")
+        return None
+
+    # fix orbital coefficients and energies for DFMP2
+    if (module_ in ['DFMP2']):
+        wfn_ = core.Wavefunction.build(wfn.molecule(), core.get_global_option('BASIS'))
+        wfn_.deep_copy(wfn)
+        refwfn = wfn.reference_wavefunction()
+        wfn_.set_reference_wavefunction(refwfn)  # refwfn not deep_copied
+        wfn_.Ca().copy(refwfn.Ca())
+        wfn_.Cb().copy(refwfn.Cb())
+        wfn_.epsilon_a().copy(refwfn.epsilon_a())
+        wfn_.epsilon_b().copy(refwfn.epsilon_b())
+        fw = core.FCHKWriter(wfn_)
+    else:
+        fw = core.FCHKWriter(wfn)
+
+    if (module_ in ['DCT', 'DFOCC']):
+        core.print_out("""FCHKWriter: Caution! For orbital-optimized correlated methods
+            the 'Orbital Energy' field contains ambiguous data.""")
+
+    # At this point we don't know the method name, so we try to search for it.
+    # idea: get the method from the variable matching closely the 'current energy'
+    varlist = core.scalar_variables()
+    current = varlist['CURRENT ENERGY']
+
+    # delete problematic entries
+    delete_list = ['CURRENT ENERGY', 'CURRENT REFERENCE ENERGY']
+    for key in delete_list[:]:
+        if key in varlist:
+            del varlist[key]
+
+    # find closest matching energy
+    for (key, val) in varlist.items():
+        if (np.isclose(val, current, 1e-10)):
+            method = key
+
+    # The labels this functions knows. Extend if needed.
+    labels = {
+        "HF": " SCF Density",
+        "SCF": " SCF Density",
+        "DFT": " SCF Density",
+        "MP2": " MP2 Density",
+        "MP3": " MP3 Density",
+        "MP4": " MP4 Density",
+        "CI": " CI Density",
+        "CC": " CC Density",
+        "DCT": " DCT Density",
+    }
+    # assign label from method name
+    fchk_label = 'UNKNOWN'
+    for key in labels:
+        if key in method:
```

Comment:
Run a `dfocc` OMP2.5 or OLCCD and see what happens. OLCCD won't be recognized, and OMP2.5 will get mistaken for OMP2. This is the big problem.

Reply:
But that's the great fun of the fchkpt format: TCTSNBN doesn't have these methods, and that's why the standard doesn't have such entries either. So, the choice is either to throw an error in such cases (comply to standard), or write something else than what the header says (status quo which the PR is trying to change)... :rofl:

Reply:
From an old forum post (http://forum.psicode.org/t/fchkwriter-writing-wrongly-formatted-fchk-files/1082) I am under the impression one needs to stick to certain labels to not break 3rd-party programs.

Thus I went with putting "MP2 Density" and "CC Density" for the cases you mention. "DCT Density" admittedly breaks with this idea. The whole fchk formatting is just somewhat.. unfortunate.

Reply:
This might be too much to ask, but how about having a "strict" `fchk` writer (default) which would strictly conform to format and always produce valid `fchk` files (ie. SCF, MP2, MP3, MP4, CC and CI, if available, only) and add a permissive tag which would print things like DCT, OMP2.5 with non-standard headings and let users handle the renaming of labels if it breaks 3rd party programs?

Reply:
That sounds like a good compromise and is a simple change.